### ISSUE TITLE
[Usage]: Logprobs Scaling with O(n) Complexity – Unexpected Performance Degradation

### ISSUE THREAD

**Issue by Rachum-thu** [vllm-project/vllm#14300]
### Your current environment

PyTorch version: 2.4.0a0+3bcc3cddb5.nv24.07
Is debug build: False
CUDA used to build PyTorch: 12.5
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.0
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.14.21-150500.55.65_13.0.73-cray_shasta_c_64k-aarch64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GH200 120GB
Nvidia driver version: 550.90.12
cuDNN version: Probably one of the following:
/usr/lib/aarch64-linux-gnu/libcudnn.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_adv.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_cnn.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_engines_precompiled.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_graph.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_heuristic.so.9.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_ops.so.9.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         aarch64
CPU op-mode(s):                       64-bit
Byte Order:                           Little Endian
CPU(s):                               288
On-line CPU(s) list:                  0-287
Vendor ID:                            ARM
Model name:                           Neoverse-V2
Model:                                0
Thread(s) per core:                   1
Core(s) per socket:                   72
Socket(s):                            4
Stepping:                             r0p0
Frequency boost:                      disabled
CPU max MHz:                          3465.0000
CPU min MHz:                          81.0000
BogoMIPS:                             2000.00
Flags:                                fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 sm3 sm4 asimddp sha512 sve asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp sve2 sveaes svepmull svebitperm svesha3 svesm4 flagm2 frint svei8mm svebf16 i8mm bf16 dgh bti
L1d cache:                            18 MiB (288 instances)
L1i cache:                            18 MiB (288 instances)
L2 cache:                             288 MiB (288 instances)
L3 cache:                             456 MiB (4 instances)
NUMA node(s):                         36
NUMA node0 CPU(s):                    0-71
NUMA node1 CPU(s):                    72-143
NUMA node2 CPU(s):                    144-215
NUMA node3 CPU(s):                    216-287
NUMA node4 CPU(s):                    
NUMA node5 CPU(s):                    
NUMA node6 CPU(s):                    
NUMA node7 CPU(s):                    
NUMA node8 CPU(s):                    
NUMA node9 CPU(s):                    
NUMA node10 CPU(s):                   
NUMA node11 CPU(s):                   
NUMA node12 CPU(s):                   
NUMA node13 CPU(s):                   
NUMA node14 CPU(s):                   
NUMA node15 CPU(s):                   
NUMA node16 CPU(s):                   
NUMA node17 CPU(s):                   
NUMA node18 CPU(s):                   
NUMA node19 CPU(s):                   
NUMA node20 CPU(s):                   
NUMA node21 CPU(s):                   
NUMA node22 CPU(s):                   
NUMA node23 CPU(s):                   
NUMA node24 CPU(s):                   
NUMA node25 CPU(s):                   
NUMA node26 CPU(s):                   
NUMA node27 CPU(s):                   
NUMA node28 CPU(s):                   
NUMA node29 CPU(s):                   
NUMA node30 CPU(s):                   
NUMA node31 CPU(s):                   
NUMA node32 CPU(s):                   
NUMA node33 CPU(s):                   
NUMA node34 CPU(s):                   
NUMA node35 CPU(s):                   
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; __user pointer sanitization
Vulnerability Spectre v2:             Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cudnn-frontend==1.5.1
[pip3] nvidia-dali-cuda120==1.39.0
[pip3] nvidia-modelopt==0.13.0
[pip3] nvidia-nvimgcodec-cu12==0.2.0.7
[pip3] nvidia-pyindex==1.0.9
[pip3] onnx==1.16.0
[pip3] optree==0.12.1
[pip3] pynvml==11.4.1
[pip3] pytorch-triton==3.0.0+989adb9a2
[pip3] pyzmq==26.0.3
[pip3] torch==2.4.0a0+3bcc3cddb5.nv24.7
[pip3] torch-tensorrt==2.5.0a0
[pip3] torchvision==0.19.0a0
[pip3] transformers==4.49.0
[conda] numpy                     2.2.3                    pypi_0    pypi
[conda] sentence-transformers     3.4.0                    pypi_0    pypi
[conda] torch                     2.5.1                    pypi_0    pypi
[conda] torchaudio                2.5.1                    pypi_0    pypi
[conda] torchvision               0.20.1                   pypi_0    pypi
[conda] transformers              4.48.1                   pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.7.4.dev142+g9804145c.d20250228
vLLM Build Flags:
CUDA Archs: 5.2 6.0 6.1 7.0 7.2 7.5 8.0 8.6 8.7 9.0+PTX; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      0-71    0               4

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NVIDIA_VISIBLE_DEVICES=all
CUBLAS_VERSION=12.5.3.2
NVIDIA_REQUIRE_CUDA=cuda>=9.0
CUDA_CACHE_DISABLE=1
TORCH_CUDA_ARCH_LIST=5.2 6.0 6.1 7.0 7.2 7.5 8.0 8.6 8.7 9.0+PTX
NCCL_VERSION=2.22.3
NVIDIA_DRIVER_CAPABILITIES=compute,utility,video
NVIDIA_PRODUCT_NAME=PyTorch
CUDA_VERSION=12.5.1.007
PYTORCH_VERSION=2.4.0a0+3bcc3cd
PYTORCH_BUILD_NUMBER=0
CUDA_VISIBLE_DEVICES=0
CUDA_VISIBLE_DEVICES=0
CUDNN_VERSION=9.2.1.18
PYTORCH_HOME=/opt/pytorch/pytorch
LD_LIBRARY_PATH=/projects/bebu/rtian2/vllm_venv/lib/python3.10/site-packages/cv2/../../lib64:/usr/local/cuda/compat/lib.real:/usr/local/cuda/compat/lib.real:/usr/local/cuda/compat/lib.real:/usr/local/lib/python3.10/dist-packages/torch/lib:/usr/local/lib/python3.10/dist-packages/torch_tensorrt/lib:/usr/local/cuda/compat/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/.singularity.d/libs
NVIDIA_BUILD_ID=100464920
CUDA_DRIVER_VERSION=555.42.06
PYTORCH_BUILD_VERSION=2.4.0a0+3bcc3cd
CUDA_HOME=/usr/local/cuda
CUDA_HOME=/usr/local/cuda
CUDA_MODULE_LOADING=LAZY
NVIDIA_REQUIRE_JETPACK_HOST_MOUNTS=
NVIDIA_PYTORCH_VERSION=24.07
TORCH_ALLOW_TF32_CUBLAS_OVERRIDE=1
CUDATOOLKIT_HOME=/opt/nvidia/hpc_sdk/Linux_aarch64/24.3/cuda/12.3
NCCL_CUMEM_ENABLE=0
TORCHINDUCTOR_COMPILE_THREADS=1

### How would you like to use vllm

**Title:** Logprobs Scaling with O(n) Complexity – Unexpected Performance Degradation  

**Description:**  
When increasing the `logprobs` parameter, I expected only a minor increase in runtime due to slicing the top-k values from the full vocabulary logits. However, my experiments show an almost O(n) increase in runtime, which suggests that retrieving logprobs is more computationally expensive than anticipated.  

### **Reproduction Code**  
```python
import time
from vllm import LLM
from vllm.sampling_params import SamplingParams

def test_generation_time(llm, logprobs_value, batch_size=32):
    sampling_params = SamplingParams(logprobs=logprobs_value, max_tokens=1)
    
    # Timed run
    start_time = time.time()
    output = llm.generate(["Tell me something about LLMs."] * batch_size,
                         sampling_params=sampling_params,
                         use_tqdm=False)
    end_time = time.time()
    
    return end_time - start_time

def main():
    print("Initializing model...")
    llm = LLM(model="Qwen/Qwen2.5-7B-Instruct", max_logprobs=152_064)  # vocab size
    
    batch_size = 32
    logprobs_values = [10, 100, 1000, 10000, 100000, 152064]
    results = []
    
    print("\nStarting tests...")
    for logprobs in logprobs_values:
        time_taken = test_generation_time(llm, logprobs, batch_size)
        results.append((logprobs, time_taken))
    
    print("\nResults Summary:")
    print("╔══════════════╦═══════════════╗")
    print("║   Logprobs   ║  Time (secs)  ║")
    print("╠══════════════╬═══════════════╣")
    for logprobs, time_taken in results:
        print(f"║ {logprobs:^12} ║ {time_taken:^13.4f} ║")
    print("╚══════════════╩═══════════════╝")

if __name__ == "__main__":
    main()
```

### **Observed Results**  
```
╔══════════════╦═══════════════╗
║   Logprobs   ║  Time (secs)  ║
╠══════════════╬═══════════════╣
║      10      ║    0.0784     ║
║     100      ║    0.0410     ║
║     1000     ║    0.1909     ║
║    10000     ║    1.9388     ║
║    100000    ║    19.9256    ║
║    152064    ║    29.2862    ║
╚══════════════╩═══════════════╝
```

### **Expected Behavior**  
Since the model inherently computes full logits for the vocabulary on every forward pass, I expected retrieving `logprobs` to involve only a minor computational overhead (e.g., sorting/selecting top-k). However, the results suggest that requesting more logprobs significantly increases runtime, implying an O(n) complexity scaling instead of an efficient selection from precomputed logits.

### **Questions:**  
1. **Why does increasing `logprobs` scale in an O(n) fashion?**  
   - Is the model recomputing or performing expensive operations instead of just slicing logits?  
2. **Is there a way to retrieve logprobs for the full vocabulary without incurring this high runtime penalty?**  
3. **Would it be possible to expose full logits instead of just logprobs?**  

### **System Info:**  
- vLLM version: 0.7.4.dev142+g9804145c.d20250228
- Model: `Qwen/Qwen2.5-7B-Instruct`  
- CUDA Version: CUDA Version: 12.5  

Looking forward to insights on whether this is expected behavior or a possible optimization opportunity! Thanks!  

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Rachum-thu** [comment]
### Update on Issue: Logprobs Scaling with O(n) Complexity – Unexpected Performance Degradation

After further investigation, I have identified a naive solution that bypasses the performance bottleneck associated with `logprobs`. Instead of directly relying on `logprobs`, I extracted the `logits` at each stage of the model execution pipeline, ensuring that they are properly propagated through all relevant function calls.

### Implementation Details:
- Modified function calls to pass and retrieve `logits` at each level:
  - **Model Execution Layer:** Updated `execute_model()` in `model_runner.py`
  - **Worker Layer:** Adjusted `execute_model()` in `worker_base.py`
  - **Executor Layer:** Ensured `execute_model()` in `executor_base.py` handles logits correctly
  - **LLM Engine Layer:** Updated `step()` in `llm_engine.py`
  - **Entrypoint Adjustments:** Made necessary changes in `_run_engine()`, `generate()`, and `beam_search()` within `llm.py`
- Introduced the following function call to obtain full logits while avoiding the `logprobs` overhead:

```python
output, logits = llm.generate(
    ["Tell me something about LLMs."] * batch_size,
    sampling_params=sampling_params,
    use_tqdm=False,
    output_logits=True
)
```

By setting `output_logits=True`, we can now retrieve the full logits directly, effectively bypassing the performance overhead introduced by `logprobs`.

This solution is now functioning correctly (at least seemingly) without causing additional complexity in execution time.

Would appreciate any feedback or suggestions from the maintainers and community!

**Comment by Rachum-thu** [comment]
### Update on Issue: Logprobs Scaling with O(n) Complexity – Unexpected Performance Degradation

After further investigation, I have identified a solution that bypasses the performance bottleneck associated with `logprobs`. Instead of directly relying on `logprobs`, I extracted the `logits` at each stage of the model execution pipeline, ensuring that they are properly propagated through all relevant function calls.

### Implementation Details:
- Modified function calls to pass and retrieve `logits` at each level:
  - **Model Execution Layer:** Updated `execute_model()` in `model_runner.py`
  - **Worker Layer:** Adjusted `execute_model()` in `worker_base.py`
  - **Executor Layer:** Ensured `execute_model()` in `executor_base.py` handles logits correctly
  - **LLM Engine Layer:** Updated `step()` in `llm_engine.py`
  - **Entrypoint Adjustments:** Made necessary changes in `_run_engine()`, `generate()`, and `beam_search()` within `llm.py`
- Introduced the following function call to obtain full logits while avoiding the `logprobs` overhead:

def test_generation_time(llm, logprobs_value, batch_size=32):
    sampling_params = SamplingParams(logprobs=logprobs_value, max_tokens=1)
    
    # Actual timed run
    start_time = time.time()
    output, logits = llm.generate(["Tell me something about LLMs."] * batch_size,
                         sampling_params=sampling_params,
                         use_tqdm=False,
                         output_logits=True)
    logprobs = compute_logprobs(logits)
    """
    (Pdb) output[0].outputs[0].logprobs
    [{444: Logprob(logprob=-0.3651916980743408, rank=1, decoded_token=' L'), 20286: Logprob(logprob=-1.8651916980743408, rank=2, decoded_token=' Large'), 11434: Logprob(logprob=-3.490191698074341, rank=3, decoded_token=' Language'), 4710: Logprob(logprob=-3.615191698074341, rank=4, decoded_token=' \n\n'), 22555: Logprob(logprob=-3.990191698074341, rank=5, decoded_token=' Sure'), 715: Logprob(logprob=-4.490191459655762, rank=6, decoded_token=' \n'), 358: Logprob(logprob=-4.615191459655762, rank=7, decoded_token=' I'), 362: Logprob(logprob=-4.740191459655762, rank=8, decoded_token=' A'), 3555: Logprob(logprob=-4.990191459655762, rank=9, decoded_token=' What'), 758: Logprob(logprob=-5.802691459655762, rank=10, decoded_token=' In')}]
    (Pdb) logprobs[0][444]
    tensor(-0.3652, device='cuda:0')
    """
    end_time = time.time()
    return end_time - start_time

By setting `output_logits=True`, we can now retrieve the full logits directly, effectively bypassing the performance overhead introduced by `logprobs`.

This solution has been tested and is now functioning correctly without causing additional complexity in execution time.

Would appreciate any feedback or suggestions from the maintainers and community!

**Comment by giacomo-brunetta** [comment]
@Rachum-thu, I was trying to do exactly what you did with logits. Your code would be very helpful to me. Is it publicly available?

**Comment by nataxcan** [comment]
I would also like this, vLLM would be terribly useful for sampling distillation data...

**Comment by Rachum-thu** [comment]
Thank you guys for looking into this! I'm connecting with the staff for the PR.
